File: src/bundler/IndexStringMap.zig
const IndexStringMap = @This();

pub const Index = bun.ast.Index;

map: std.AutoArrayHashMapUnmanaged(Index.Int, []const u8) = .{},

pub fn deinit(self: *IndexStringMap, allocator: std.mem.Allocator) void {
    for (self.map.values()) |value| {
        allocator.free(value);
    }
    self.map.deinit(allocator);
}

pub fn get(self: *const IndexStringMap, index: Index.Int) ?[]const u8 {
    return self.map.get(index);
}

pub fn put(self: *IndexStringMap, allocator: std.mem.Allocator, index: Index.Int, value: []const u8) !void {
    const duped = try allocator.dupe(u8, value);
    errdefer allocator.free(duped);
    try self.map.put(allocator, index, duped);
}

const bun = @import("bun");
const std = @import("std");

File: src/bundler/IndexStringMap.rs
use bun_collections::ArrayHashMap;

/// The underlying integer repr of `Index`.
pub(crate) use crate::IndexInt;

#[derive(Default)]
pub struct IndexStringMap {
    map: ArrayHashMap<IndexInt, Box<[u8]>>,
}

impl IndexStringMap {
    pub(crate) fn get(&self, index: IndexInt) -> Option<&[u8]> {
        self.map.get(&index).map(|v| v.as_ref())
    }

    pub(crate) fn put(
        &mut self,
        index: IndexInt,
        value: impl AsRef<[u8]>,
    ) -> Result<(), bun_alloc::AllocError> {
        let duped = Box::<[u8]>::from(value.as_ref());
        // errdefer arena.free(duped) — deleted: `duped` is Drop, `?` handles cleanup.
        self.map.insert(index, duped);
        Ok(())
    }
}
